uvicorn/protocols/websockets/websockets_impl.py

```
@@ -278,43 +282,83 @@ async def asgi_send(self, message: "ASGISendEvent") -> None:
+            if self.response_body is None:
+                if message_type == "websocket.accept":
+                    message = cast("WebSocketAcceptEvent", message)
+                    self.logger.info(
+                        '%s - "WebSocket %s" [accepted]',
+                        self.scope["client"],
+                        get_path_with_query_string(self.scope),
+                    )
+                    self.initial_response = None
+                    self.accepted_subprotocol = cast(
+                        Optional[Subprotocol], message.get("subprotocol")
+                    )
+                    if "headers" in message:
+                        self.extra_headers.extend(
+                            # ASGI spec requires bytes
+                            # But for compatibility we need to convert it to strings
+                            (name.decode("latin-1"), value.decode("latin-1"))
+                            for name, value in message["headers"]
+                        )
+                    self.handshake_started_event.set()
+
+                elif message_type == "websocket.close":
+                    message = cast("WebSocketCloseEvent", message)
+                    self.logger.info(
+                        '%s - "WebSocket %s" 403',
+                        self.scope["client"],
+                        get_path_with_query_string(self.scope),
+                    )
+                    self.initial_response = (http.HTTPStatus.FORBIDDEN, [], b"")
+                    self.handshake_started_event.set()
+                    self.closed_event.set()
+
+                elif message_type == "websocket.http.response.start":
+                    message = cast("WebSocketResponseStartEvent", message)
+                    try:
+                        status = http.HTTPStatus(message["status"])
+                    except AttributeError:
+                        status = http.HTTPStatus.FORBIDDEN
```

By the spec, `status` should always be provided.
```suggestion
```

---

Ah, it is an `AttributeError` which happens if the lookup from `http.HTTPStatus` fails.  That is, if the use has provided
a code which isn't part of the enum.  This needs to be an enum, otherwise sending the response later will quietly fail.
I guess we could just allow the lookup here to fail, I was trying to be defensive.  Do you agree, or should we fallback to this code if an invalid code is provided?

---

> This needs to be an enum

Does `websockets` package "doesn't work" if you provide the int `message["status"]`?

---

yes, see below.

---

Let's not silently replace it, tho... Let's add a log message here, please :+1:

---

Adding a test for this case (trying to send a message with invalid status) showed that the wsproto will cause a 500 error to be raised.  So I think it is fine if we also get that error for websockets.  I'm removing the check and adding a test to verify that the server produces a 500 error.

---

I'm fine with this compromise. Thanks :+1: